File: 216-combination-sum-iii/combination-sum-iii.js
/**
 * @param {number} k
 * @param {number} n
 * @return {number[][]}
 */
var combinationSum3 = function(k, n) {
    let arr = []
    
    let bt = (i, nums, sum) => {
        if (sum > n || nums.length > k) return

        if (nums.length == k && sum == n) {
            arr.push([...nums])
            return
        }

        for (; i < 10; i++) {
            nums.push(i)
            bt(i + 1, nums, sum + i)
            nums.pop()
        }
    }
    bt(1, [], 0)

    return arr
};

File: 216-combination-sum-iii/combination-sum-iii.py
class Solution:
    def combinationSum3(self, k: int, n: int) -> List[List[int]]:
        arr = []

        def bt(i, nums, s):
            if s > n or len(nums) > k:
                return
            
            if s == n and len(nums) == k:
                arr.append(nums.copy())
                return

            for x in range(i, 10):
                nums.append(x)
                bt(x + 1, nums, s + x)
                nums.pop()
        
        bt(1, [], 0)
        return arr